File: frontend/src/lib/folderAccess.ts
/** The one sentence we say when the operating system, not the project, is in the way.
 *
 * On macOS the first read of a folder under Documents triggers a permission dialog,
 * and until it is answered the read simply *blocks* — no error, no timeout. If the
 * dialog is hidden behind a notification, the app spins "Enumerating files…" forever.
 * Ten files, seven minutes, no explanation. The fix is one click the person never saw,
 * so the app's job is to point at it.
 *
 * Kept here, not inline in a component, because the scan is not the only thing that
 * reads the folder: indexing, Ask and the Investigator hit the same wall if the grant
 * is revoked on a live workspace. Must stay in step with the backend's
 * `core/domain/folder_access.py` — the backend sends this same sentence as the failure
 * message, and the watchdog below shows it before any failure arrives.
 */
export const FOLDER_PERMISSION_HINT =
  "macOS may be asking for permission to read this folder — look for a system dialog. " +
  "You can also grant access in System Settings → Privacy & Security → Files and Folders.";

/** How long a job may make no progress at all before we suspect the dialog.
 *
 * A healthy walk reports its running file count as it descends, so silence means
 * stuck, not slow. Five seconds is long enough not to cry wolf on a cold disk. */
export const NO_PROGRESS_HINT_AFTER_MS = 5000;

/** Is this job actually one that could be waiting on the folder dialog?
 *
 * Only the scan, and only before it has counted its first file: macOS asks on the
 * *first* read of the folder, so a walk that has reported any count is already in.
 * Indexing is past that gate — it embeds text we have already read — and a pause
 * there is a large batch, not a permission problem. The hint once appeared over an
 * indexing bar moving through 65%, which is the worst kind of message: confidently
 * wrong about the very thing the person is looking at. */
export function couldBeWaitingForFolderPermission(job: {
  kind: "scan" | "index";
  current: number | null;
}): boolean {
  // `null` = the job has not reported anything yet; 0 = it reported, and found
  // nothing so far. Both mean "no file has been counted", which is the only state
  // the dialog can be blocking.
  return job.kind === "scan" && !job.current;
}

File: backend/app/core/domain/folder_access.py
"""When the operating system, not the project, is the problem.

On macOS the first read of a folder under Documents/Desktop/Downloads triggers a TCC
prompt. Until the person answers it, the very first `listdir` simply *blocks* — no
error, no timeout. If the dialog is hidden behind a notification, the scan sits there
forever and the app cheerfully spins "Enumerating files…". Ten files, seven minutes,
no explanation. That is the worst kind of failure: the app looks broken, and the fix
is one click the person never saw.

If they deny it (or the grant is later revoked), we get an EACCES/EPERM instead —
and today `os.walk` swallows it, so the scan "succeeds" with a silently smaller file
set. A quietly incomplete index is worse than an honest error.

One sentence, in one place, used by the scan, the index and anything else that reads
the folder — and mirrored in the frontend (see `lib/folderAccess.ts`). It names the
cause and the exact place to fix it, because "permission denied" helps nobody.
"""

from __future__ import annotations

import errno

FOLDER_PERMISSION_MESSAGE = (
    "This app doesn't have permission to read that folder. macOS may be asking for "
    "access right now — look for a system dialog. You can also grant it in System "
    "Settings → Privacy & Security → Files and Folders."
)


class FolderPermissionError(PermissionError):
    """The folder exists, but the OS will not let us read it.

    Carries the user-facing sentence as its message, so every layer above — the job
    runner, the API, the UI — can show it without re-deriving it from an errno.
    """

    def __init__(self, path: str) -> None:
        super().__init__(FOLDER_PERMISSION_MESSAGE)
        self.path = path


def is_permission_error(error: OSError) -> bool:
    """EACCES/EPERM, however the platform spells it today."""
    return isinstance(error, PermissionError) or error.errno in (errno.EACCES, errno.EPERM)
